### ISSUE TITLE
[BUG] Stage2 would get faulty gradient if parameters on different ranks get their gradients on different order.

### ISSUE THREAD

**Issue by li-yi-dong** [microsoft/DeepSpeed#2343]
**Describe the bug**
Stage2 would get faulty gradient if parameters on different ranks get their gradients on different order.

**To Reproduce**
Here is the code:
```Python
import deepspeed
import torch
import torch.distributed as dist
from torch.optim import Optimizer

import argparse


parser = argparse.ArgumentParser('Training and evaluation script', add_help=False)
parser.add_argument("--deepspeed", action="store_true", help="toggle deepspeed")
parser.add_argument("--deepspeed_config", type=str, default="", help="path to deepspeed config")
args, _ = parser.parse_known_args()


deepspeed.init_distributed()


class M(torch.nn.Module):
    def __init__(self):
        super().__init__()

        self.layer1 = torch.nn.Parameter(torch.Tensor([1, 1, 1, 1]))
        self.layer2 = torch.nn.Parameter(torch.Tensor([2, 2 ,2, 2]))

    def forward(self, input1, input2):
        if dist.get_rank() == 0:
            a = input1 * self.layer1
            b = input2 * self.layer2
        else:
            b = input2 * self.layer2
            a = input1 * self.layer1

        c = a + b
        return c.sum()


class Opt(Optimizer):
    def __init__(self, params):
        defaults = dict()
        super(Opt, self).__init__(params, defaults)

    def step(self):
        for group in self.param_groups:
            for p in group['params']:
                print('In optim', p.grad, flush=True)


device = torch.device('cuda:0')

input1 = torch.Tensor([1]).to(device=device)
input2 = torch.Tensor([2]).to(device=device)

model = M()
model.to(device=device)

opt = Opt(params=model.parameters())
model, opt, _, _ = deepspeed.initialize(args=args, model=model, optimizer=opt)
loss = model(input1, input2)
model.backward(loss)
model.step()
```

**Expected behavior**
The gradient printed by the dummy `Optimizer` should be `tensor([1., 1., 1., 1.], device='cuda:0')` or `tensor([2., 2., 2., 2.], device='cuda:0')`.
Yet, it comes out 
<img width="1525" alt="截屏2022-09-22 上午10 31 43" src="https://user-images.githubusercontent.com/73142299/191645299-9ff37aea-e30f-4823-94e4-263a0cde4b9a.png">

**ds_report output**
<img width="1206" alt="截屏2022-09-22 上午10 47 35" src="https://user-images.githubusercontent.com/73142299/191647132-27ddff54-5b4c-4e47-9fa2-4f45afbe52ca.png">

**Launcher context**
ds_config.json
```JSON
{
  "train_micro_batch_size_per_gpu": 40,
  "train_batch_size": 80,
  "prescale_gradients": false,
  "sparse_gradients": false,
  "fp16": {
    "enabled": false,
    "auto_cast": false,
    "loss_scale": 0,
    "initial_scale_power": 32,
    "loss_scale_window": 1000,
    "hysteresis": 2,
    "min_loss_scale": 1
  },
  "zero_optimization": {
    "stage": 2,
    "allgather_partitions": true,
    "allgather_bucket_size": 2e9,
    "overlap_comm": true,
    "reduce_scatter": true,
    "reduce_bucket_size": 4,
    "contiguous_gradients": false
  },
  "steps_per_print": 1,
  "wall_clock_breakdown": true,
  "dump_state": false,
  "autotuning": {
    "enabled": false,
    "results_dir": null,
    "exps_dir": null,
    "overwrite": false,
    "metric": "throughput",
    "start_profile_step": 3,
    "end_profile_step": 5,
    "fast": true,
    "max_train_batch_size": null,
    "mp_size": 1,
    "num_tuning_micro_batch_sizes": 3,
    "tuner_type": "model_based",
    "tuner_early_stopping": 5,
    "tuner_num_trials": 50,
    "arg_mappings": null
  },
  "flops_profiler": {
    "enabled": false,
    "profile_step": 20,
    "module_depth": -1,
    "top_modules": 1,
    "detailed": true,
    "output_file": null
  },
  "zero_allow_untested_optimizer": true
}
```

**Additional context**
I discover this bug while I walk through the source code in `deepspeed/runtime/zero/stage_1_and_2.py`.
`reduce_independent_p_g_buckets_and_remove_grads`, of which would be registered to parameters' gradient accumulator, would trigger all_reduce once the pending gradients exceeded `reduce_bucket_size`, no matter the order this hook been triggered.
```Python
def reduce_independent_p_g_buckets_and_remove_grads(self, param, i):
    if self.elements_in_ipg_bucket + param.numel() > self.reduce_bucket_size:
        self.report_ipg_memory_usage("In ipg_remove_grads before reduce_ipg_grads", param.numel())
        self.reduce_ipg_grads()
```
Instead, `torch.nn.parallel.DistributedDataParallel` would assign each parameter to a bucket in advance and all_reduced the predefined buckets according to buckets' order restrictively.
```C
// Called when the bucket at the specified index is ready to be reduced.
void Reducer::mark_bucket_ready(size_t bucket_index) {
  TORCH_INTERNAL_ASSERT(bucket_index >= next_bucket_);

  // Buckets are reduced in sequence. Ignore this bucket if
  // it's not its turn to be reduced.
  if (bucket_index > next_bucket_) {
    return;
  }

  // Keep going, until we either:
  // - have kicked off reduction for all buckets, or
  // - found a bucket that's not yet ready for reduction.
  for (; next_bucket_ < buckets_.size() && buckets_[next_bucket_].pending == 0;
       next_bucket_++) {
    num_buckets_ready_++;
    if (num_buckets_ready_ == 1 && should_collect_runtime_stats()) {
      record_backward_comm_start_time();
    }
    auto& bucket = buckets_[next_bucket_];
    all_reduce_bucket(bucket);
  }
}
```

**Comment by jomayeri** [comment]
@li-yi-dong I was unable to produce the same gradient results when running your example code. Even running multiple times in a loop produced consistent results.
![image](https://user-images.githubusercontent.com/114769929/197880469-e806a3ae-3b38-4fec-9a23-a7adccdf8e4c.png)

Is there any more information or code you could provide for reproduction?

**Comment by li-yi-dong** [comment]
> @li-yi-dong I was unable to produce the same gradient results when running your example code. Even running multiple times in a loop produced consistent results. ![image](https://user-images.githubusercontent.com/114769929/197880469-e806a3ae-3b38-4fec-9a23-a7adccdf8e4c.png)
> 
> Is there any more information or code you could provide for reproduction?

It seems that you was running on a signal node, so that the `Optimizer` saw all 8 gradients. 
You should run with 2 nodes and set `"reduce_bucket_size": 4`, letting `stage2` split gradients and optimizer states and `all_reduce` gradients between the 2 nodes.

**Comment by jomayeri** [comment]
I was able to reproduce the discrepancy for ZeRO stage 2 across ranks. The reason for the different Optimizer results is clear from the forward pass in your example code. ZeRO stage 2 operates under the assumption that all ranks are performing the same job in the same order. This is an assumption that allows for added memory savings in Stage 2. Your forward pass is modifying the order of operations based on the rank, which is why you are seeing different results.

I also confirmed that ZeRO Stage 1 would get you the expected result, so that could be an alternative if you still need the ranks to have a different ordering in the model.

**Comment by li-yi-dong** [comment]
> I was able to reproduce the discrepancy for ZeRO stage 2 across ranks. The reason for the different Optimizer results is clear from the forward pass in your example code. ZeRO stage 2 operates under the assumption that all ranks are performing the same job in the same order. This is an assumption that allows for added memory savings in Stage 2. Your forward pass is modifying the order of operations based on the rank, which is why you are seeing different results.
> 
> I also confirmed that ZeRO Stage 1 would get you the expected result, so that could be an alternative if you still need the ranks to have a different ordering in the model.

Hi @jomayeri 
I don't see any notification related to this in DeepSpeed's user manuals. If this is intended behavior, at least, it should be warned to the users. Otherwise, it is almost impossible to root cause when training failed 🙃
